lib/src/device_authorization_grant.dart

```diff
@@ -0,0 +1,385 @@
+///
+/// With this method of authorization, the device requests a device code.
+/// Afterwards, a user must authorize the device. The device being authorized
+/// attempts to retrieve tokens at a defined interval. This is used to obtain
+/// [Credentials] and create a fully authorized [Client].
+///
+/// To use this class, you must first call [getDeviceCode] to get the user code
+/// and confirmation url, which the resource owner can then use to confirm the
+/// device.The device polls the authorization server for an access token at a
+/// defined interval using [pollForToken], and creates a [Client] if successful.
+///
+/// [device code grant]: https://tools.ietf.org/html/rfc8628#section-1
+class DeviceAuthorizationGrant {
+  /// The function used to parse parameters from a host's response.
+  final GetParameters _getParameters;
+
+  /// The client identifier for this client.
+  ///
+  /// The authorization server will issue each client a separate client
+  /// identifier and secret, which allows the server to tell which client is
+  /// accessing it. Some servers may also have an anonymous identifier/secret
+  /// pair that any client may use.
+  ///
+  /// This is usually global to the program using this library.
+  final String identifier;
+
+  /// The client secret for this client.
+  ///
+  /// The authorization server will issue each client a separate client
+  /// identifier and secret, which allows the server to tell which client is
+  /// accessing it. Some servers may also have an anonymous identifier/secret
+  /// pair that any client may use.
+  ///
+  /// This is usually global to the program using this library.
+  ///
+  /// Note that clients whose source code or binary executable is readily
+  /// available may not be able to make sure the client secret is kept a secret.
+  /// This is fine; OAuth2 servers generally won't rely on knowing with
+  /// certainty that a client is who it claims to be.
+  final String? secret;
+
+  /// A URL provided by the authorization server that this library uses to
+  /// obtain a device_code.
+  ///
+  /// This will usually be listed in the authorization server's OAuth2 API
+  /// documentation.
+  final Uri deviceEndpoint;
+
+  /// A URL provided by the authorization server that this library uses to
+  /// obtain long-lasting credentials.
+  ///
+  /// This will usually be listed in the authorization server's OAuth2 API
+  /// documentation.
+  final Uri tokenEndpoint;
+
+  /// Callback to be invoked whenever the credentials are refreshed.
+  ///
+  /// This will be passed as-is to the constructed [Client].
+  final CredentialsRefreshedCallback? _onCredentialsRefreshed;
+
+  /// Whether to use HTTP Basic authentication for authorizing the client.
+  final bool _basicAuth;
+
+  /// A [String] used to separate scopes; defaults to `" "`.
+  final String _delimiter;
+
+  /// The HTTP client used to make HTTP requests.
+  http.Client? _httpClient;
```

Comment:
Why is this nullable? the constructor doesn't seem to allow it.

Reply:
fixed by commit [168a984](https://github.com/dart-lang/oauth2/pull/112/commits/168a9847c5ede4e4a335846e65ae70cd782ae302)